### PR TITLE
[OKR] SQL Handbook

### PR DESCRIPTION
**Pull request by PetarCurkovic**
Task [#415](https://app.productive.io/1-infinum/task/2044837)

#### Aim

Add list of SQL query examples for our SQL catalogue, with some interesting SQL topics.

#### Solution

* Added query examples for SQL handbook
* Added General info, Tips&Tricks, and Indexes chapters

### CODE REVIEW

File: Databases/SQL/Tips & Tricks.md [hunk 1]
```diff
@@ -0,0 +1,47 @@
...
+
+## UNION (ALL)
+
+When you use `UNION` operator for combining result sets of two or more query statements, double-check if you would use `UNION` or `UNION ALL`.
+
+The `UNION` operator is *less performant* because it **removes all duplicate rows** from the combined data set.
+To retain the duplicate rows, you use the `UNION ALL` instead.
+
+
+## VIEWs
```
Comment (nikajukic) [inline Databases/SQL/Tips & Tricks.md:16]: I'd maybe mention few more things here:

- you can use a view as an ActiveRecord model (with example)
- scenic gem for managing views?
- are materialized views a way to go if you have a lot of inserts (you'd need to refresh it constantly)
- tips on when to use which view?

**Review by nikajukic — COMMENTED** [review]
Great job @PetarCurkovic!! 🎉 👌

**Comment by Benaaaaa** [comment]
Hey, just a heads up, in both **indexes** and **tips & tricks** chapters in the beta version of the [handbook](https://beta.infinum.com/handbook/rails/databases/sql/tips-and-tricks) the bullet points are not being rendered properly. I tried investigating why is that but don't know much markdown and it doesn't make sense at the moment why is it like that

**Comment by PetarCurkovic** [comment]
> Hey, just a heads up, in both **indexes** and **tips & tricks** chapters in the beta version of the [handbook](https://beta.infinum.com/handbook/rails/databases/sql/tips-and-tricks) the bullet points are not being rendered properly. I tried investigating why is that but don't know much markdown and it doesn't make sense at the moment why is it like that

@Benaaaaa thx, I already fixed that (will be pushed soon)!
There is a missing new line between the text and bullet list.

File: Databases/SQL/General.md [hunk 2]
```diff
@@ -0,0 +1,64 @@
+## Intro
+
+When you work on backend systems, you will always have some kind of data storage where you will persist the application data.
```
Comment (cilim) [inline Databases/SQL/General.md:3]:
```suggestion
Working on backend systems almost always necessitates the use of a database where application data can be stored.
```

File: Databases/SQL/General.md [hunk 3]
```diff
@@ -0,0 +1,64 @@
...
+The majority of our applications, use a relational database as a data storage solution.
```
Comment (cilim) [inline Databases/SQL/General.md:4]:
```suggestion
The majority of our applications use a relational database as a data storage solution.
```

File: Databases/SQL/General.md [hunk 4]
```diff
@@ -0,0 +1,64 @@
...
+
+To be able to communicate with your database, you need to use SQL.
```
Comment (cilim) [inline Databases/SQL/General.md:6]:
```suggestion
SQL (structured query language) is the tool for communicating with the database.
```

File: Databases/SQL/General.md [hunk 5]
```diff
@@ -0,0 +1,64 @@
...
+We usually leave that communication to our ORM (`ActiveRecord`). But sometimes, ORM is not powerful enough for a specific problem and you must write raw SQL.
+
+To be able to **solve complex problems** and to get **better understanding** of how the ORM works, here are some query examples and advice to help you with that!
```
Comment (cilim) [inline Databases/SQL/General.md:9]:
```suggestion
To be able to **solve complex problems** and to get **better understanding** of how the ORM works, here are some query examples and advices to help you with that!
```

File: Databases/SQL/General.md [hunk 6]
```diff
@@ -0,0 +1,64 @@
...
+
+All query examples will be compatible with **PostgreSQL standard**.
+
+SQL keywords are formatted in all capitals to make them stand out from the text, as in **SELECT**.
+
+SQL tables and columns are spelled in lowercase, and words are separated by underscores, as in `users`, `first_name`.
+
+It would make your life easier if you align and format a query you currently working on. It would be easier to debug and identify parts of the query.
+
+SQL is correctly pronounced "ess-cue-ell", not "see-quell".
```
Comment (cilim) [inline Databases/SQL/General.md:22]: 👏 👏 👏

File: Databases/SQL/General.md [hunk 7]
```diff
@@ -0,0 +1,64 @@
...
+
+We set up a PostgreSQL database instance that you can use to test your queries.
+The credentials are stored in our **1Password**.
+
+
+## DB GUI Clients
+
+There are plenty of GUI tools for interacting with PostgreSQL. Here is the list of the most popular GUI clients - feel free to choose the one you like:
+
+  * [Postico](https://eggerapps.at/postico/) (only for Mac - we have a full version license)
```
Comment (cilim) [inline Databases/SQL/General.md:37]:
```suggestion
  * [Postico](https://eggerapps.at/postico/) (only for Mac - we have a full version license that you can acquire from your team lead)
```

File: Databases/SQL/Indexes.md [hunk 8]
```diff
@@ -0,0 +1,61 @@
...
+
+## Data Structure
+
+The data structure that is used for indexes is a **B-tree** (balanced tree). That's because the tree depth is equal for all nodes (the distance between root and leaf nodes).
+
+## Why a query does not use an index?
+
+The execution plan of a query can show that the database does not use the index, and uses **sequential scan**.
+Possible situations:
+  * majority of rows are getting fetched as part of the SQL query
```
Comment (cilim) [inline Databases/SQL/Indexes.md:11]:
```suggestion
Possible situations:
  
  * majority of rows are getting fetched as part of the SQL query
```

I believe a new row is missing here for bullet formatting

File: Databases/SQL/Tips & Tricks.md [hunk 9]
```diff
@@ -0,0 +1,48 @@
...
+	* reducing complexity by naming a part of the complex query
+  * exposing data subset to the outer world
+
+Plain `VIEW` contains only a definition - that means the query would be executed each time you fetch data from the view.
+
+The results can be "cached" by using `MATERIALIZED VIEW`. Basically, `MATERIALIZED VIEW` acts as a read-only plain table. To update data in materialized views, use `REFRESH` statement.
+
+`REFRESH` completely replaces the content of a materialized view. That means the table is locked when the refresh process is running. To refresh a materialized view without locking, use `REFRESH MATERIALIZED VIEW CONCURRENTLY`.
+
+> You can add indexes to materialized views!
```
Comment (cilim) [inline Databases/SQL/Tips & Tricks.md:29]:
```suggestion
> You can add indexes to materialized views, but not plain database views.
```

File: Databases/SQL/Tips & Tricks.md [hunk 10]
```diff
@@ -0,0 +1,48 @@
...
+  * `ST_Point`
+  * `ST_ClosestPoint`
+  * `ST_Contains`
+  * `ST_Covers`
+  * `ST_Collect`
+  * `ST_Centroid`
+  * `ST_DWithin`
+  * `ST_ClusterDBSCAN`
+
+But, sometimes, PostGIS might be overkill for your use case - read more here [why you (probably) don't need PostGIS](https://blog.rebased.pl/2020/04/07/why-you-probably-dont-need-postgis.html).
```
Comment (cilim) [inline Databases/SQL/Tips & Tricks.md:48]:
```suggestion
PostGIS might be overkill in some scenarios - read more here [why you (probably) don't need PostGIS](https://blog.rebased.pl/2020/04/07/why-you-probably-dont-need-postgis.html).
```

**Review by cilim — COMMENTED** [review]
Great work! 👏 🔥 🙂

**Comment by Benaaaaa** [comment]
One proposal on the query examples: the SQL solution process is never top-down and so just reading the solution won't give us much info on how to maybe approach some problems. If it is possible could there be added a Solving steps section on what things were first solved and how the solution came to be? Sort of like you did it in the backend talks. For example just reading the solutions that are using the window function and PG documentation I am still having a hard time understanding how it exactly works, but when it was introduced on how it worked in the backend talks, it was much more clearer.

File: Databases/SQL/Query Examples/05 - CSV.md [hunk 11]
```diff
@@ -0,0 +1,64 @@
+## Problem (1)
+
+You have to export users with surname “Smith” to CSV.
+
+
+## Problem (2)
+
+You have to import new users from large CSV file.
```
Comment (cukoreva) [inline Databases/SQL/Query Examples/05 - CSV.md:8]:
```suggestion
You have to import new users from a large CSV file.
```

File: Databases/SQL/Query Examples/05 - CSV.md [hunk 12]
```diff
@@ -0,0 +1,64 @@
+## Problem (1)
+
+You have to export users with surname “Smith” to CSV.
```
Comment (cukoreva) [inline Databases/SQL/Query Examples/05 - CSV.md:3]:
```suggestion
You have to export users with the surname “Smith” to CSV.
```

File: Databases/SQL/Query Examples/08 - UPDATE with subquery.md [hunk 13]
```diff
@@ -0,0 +1,109 @@
+## Problem
+
+For the posts table, we want to separate category to its own table and then reference category by the foreign key.
+Write a query that will update current posts table data to the suggested structure.
```
Comment (cukoreva) [inline Databases/SQL/Query Examples/08 - UPDATE with subquery.md:4]:
```suggestion
Write a query that will update the current posts table data to the suggested structure.
```

File: Databases/SQL/Tips & Tricks.md [hunk 14]
```diff
@@ -0,0 +1,48 @@
+## N-queries
+
+You should avoid multiple DB queries from the application if you can fetch needed data in a single query. Always check what queries are executed in your (Rails) log.
```
Comment (cukoreva) [inline Databases/SQL/Tips & Tricks.md:3]:
```suggestion
You should avoid multiple DB queries from the application if you can fetch the needed data in a single query. Always check what queries are executed in your (Rails) log.
```

File: Databases/SQL/Tips & Tricks.md [hunk 15]
```diff
@@ -0,0 +1,48 @@
...
+
+When you use `UNION` operator for combining result sets of two or more query statements, double-check if you would use `UNION` or `UNION ALL`.
+
+The `UNION` operator is *less performant* because it **removes all duplicate rows** from the combined data set.
+To retain the duplicate rows, you use the `UNION ALL` instead.
+
+
+## VIEWs
+
+Database `VIEW` is a pretty useful SQL mechanism. You can use it for:
```
Comment (cukoreva) [inline Databases/SQL/Tips & Tricks.md:18]:
```suggestion
A database `VIEW` is a pretty useful SQL mechanism. You can use it for:
```

File: Databases/SQL/Tips & Tricks.md [hunk 16]
```diff
@@ -0,0 +1,48 @@
...
+  * hiding complexity of queries
```
Comment (cukoreva) [inline Databases/SQL/Tips & Tricks.md:19]:
```suggestion
  * hiding the complexity of queries
```

File: Databases/SQL/Tips & Tricks.md [hunk 17]
```diff
@@ -0,0 +1,48 @@
...
+	* reducing complexity by naming a part of the complex query
```
Comment (cukoreva) [inline Databases/SQL/Tips & Tricks.md:20]:
```suggestion
	* reducing the complexity by naming a part of the complex query
```

File: Databases/SQL/Tips & Tricks.md [hunk 18]
```diff
@@ -0,0 +1,48 @@
...
+  * exposing data subset to the outer world
```
Comment (cukoreva) [inline Databases/SQL/Tips & Tricks.md:21]:
```suggestion
  * exposing a data subset to the outer world
```

File: Databases/SQL/Query Examples/09 - LATERAL JOIN.md [hunk 19]
```diff
@@ -0,0 +1,278 @@
+## Problem
+
+Prepare a database view which will hold the data for similar posts.
+For calculating the similarity score, use next formula:
```
Comment (cukoreva) [inline Databases/SQL/Query Examples/09 - LATERAL JOIN.md:4]:
```suggestion
For calculating the similarity score, use the following formula:
```

File: Databases/SQL/Tips & Tricks.md [hunk 20]
```diff
@@ -0,0 +1,48 @@
...
+
+Plain `VIEW` contains only a definition - that means the query would be executed each time you fetch data from the view.
```
Comment (cukoreva) [inline Databases/SQL/Tips & Tricks.md:23]:
```suggestion
A plain `VIEW` contains only a definition - that means the query will be executed each time you fetch data from the view.
```

File: Databases/SQL/Tips & Tricks.md [hunk 21]
```diff
@@ -0,0 +1,48 @@
...
+
+The results can be "cached" by using `MATERIALIZED VIEW`. Basically, `MATERIALIZED VIEW` acts as a read-only plain table. To update data in materialized views, use `REFRESH` statement.
```
Comment (cukoreva) [inline Databases/SQL/Tips & Tricks.md:25]:
```suggestion
The results can be "cached" by using a `MATERIALIZED VIEW`. Basically, a `MATERIALIZED VIEW` acts as a read-only plain table. To update data in materialized views, use the `REFRESH` statement.
```

File: Databases/SQL/Tips & Tricks.md [hunk 22]
```diff
@@ -0,0 +1,48 @@
...
+
+> You can add indexes to materialized views!
+
+
+## PostGIS
+
+When you want to work with geographical data, most likely you'll end up using [PostGIS](https://postgis.net/). PostGIS is an extension for the PostgreSQL database.
+
+In the official [documentation](https://postgis.net/docs/), you can find more information about data types, functions, and operators.
+Here's the list of some useful functions and geometry constructors, which can be your starting point when you'll start playing with PostGIS:
```
Comment (cukoreva) [inline Databases/SQL/Tips & Tricks.md:37]:
```suggestion
Here's the list of some useful functions and geometry constructors, which can be your starting point when you start playing with PostGIS:
```

**Review by cukoreva — CHANGES_REQUESTED** [review]

**Review by JureCindro — APPROVED** [review]

**Review by uncoverd — APPROVED** [review]
I gave this a brief look and I like what I'm seeing 💪 We have a lay of the land and a sizable collection of examples already and I'm guessing more will be added as we find them.

Let's make sure we keep periodically and proactively pinging the team to add interesting examples here 🙏 

Good job @PetarCurkovic, you reduced your job security 😄

**Review by nikajukic — APPROVED** [review]

**Review by cilim — APPROVED** [review]

**Review by cukoreva — APPROVED** [review]

**Pull request merged**